### PR TITLE
omit side effects in unreachable modules (#551)

### PR DESCRIPTION
**Pull request by calvinmetcalf**
the idea is that if you don't actually import anything from a module (i.e. because everything that imported it has been shaken out of the tree) then don't include it's side effects.

This would be very helpful for situations where you have something like a lodash or other utility thing that people are going to only one one or 2 functions out of.  So if you want `_.pluck` it shouldn't matter to you that `_.somethignWeird` depends on `giantModuleWithSideEffects`.

### CODE REVIEW

**Comment by eventualbuddha** [comment]
This seems reasonable to me. Thanks!

**Comment by Rich-Harris** [comment]
Thanks @calvinmetcalf. I'm not sure if this complies with the spec though – wouldn't a native module loader execute that code? Totally understand the motivation for omitting it but it's important to adhere to the proper semantics

**Comment by calvinmetcalf** [comment]
a bundler and a loader ~~are~~ should never going to have exactly the same semantics because they are trying to do slightly different things.  Take browserify which statically resolves all dependencies without execution which does deviate from how node module loader works but in a way that makes more packages work most of the time.

Having a pedantic mode or no tree shaking mode might be a helpful addition for the minority of cases where it needs to act exactly like the module spec.

The practical backround was trying to get `import {randomBytes}  from 'crypto'` to work which ended up bringing in almost all of the dependencies, I'd guess in large part due to the inherits pattern which likely is incorrectly flagged as a side effect.

**Comment by wearhere** [comment]
Since @calvinmetcalf mentioned this in #552, what's the status of this?

I agree with "a bundler and a loader should never have exactly the same semantics because they are trying to do slightly different things" fwiw. It seems very much in line with the vision of tree shaking that you could shake side effects out along with the module that defined them i.e. if all imports had been eliminated.

Notionally wouldn't the side effects be relevant only to the module that did the side effects, anyway? So if the entry point didn't end up using anything from the module, then it wouldn't care about the side effects either.

Perhaps this wouldn't be true all the time but in that case there could be a "pedantic or no tree shaking mode" like @calvinmetcalf suggests.

**Comment by calvinmetcalf** [comment]
here are some real life pieces of code this hit this issue, this code from the browserify buffer shim causes the buffer code and all of it's deps from never being treeshaken out problematic line is highlighted


```js
if (Buffer.TYPED_ARRAY_SUPPORT) {
  Buffer.prototype.__proto__ = Uint8Array.prototype
  Buffer.__proto__ = Uint8Array
  if (typeof Symbol !== 'undefined' && Symbol.species &&
      Buffer[Symbol.species] === Buffer) {
    // Fix subarray() in ES2016. See: https://github.com/feross/buffer/pull/97
//VVVVVVVVV
    Object.defineProperty(Buffer, Symbol.species, {
      value: null,
      configurable: true
    })
//^^^^^^^^^^^
  }
}
```

the above one is relatively straightforward why it's causing the issue rollup is being conservative about the object assign even though buffer otherwise would be tree shaken out

```js
Buffer.TYPED_ARRAY_SUPPORT = global.TYPED_ARRAY_SUPPORT !== undefined
  ? global.TYPED_ARRAY_SUPPORT
  : typedArraySupport()
function typedArraySupport () {
  try {
    var arr = new Uint8Array(1)
    arr.__proto__ = {__proto__: Uint8Array.prototype, foo: function () { return 42 }}
// VVVVVVVV
    return arr.foo() === 42 && // typed array instances can be augmented
        typeof arr.subarray === 'function' && // chrome 9-10 lack `subarray`
        arr.subarray(1, 1).byteLength === 0 // ie10 has broken `subarray`
// ^^^^^^^^^^^^^
  } catch (e) {
    return false
  }
}
```

why the above code causes a tree shaking issue is beyond me, then from a dep of buffer

```js
function init () {
  var code = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/'
  for (var i = 0, len = code.length; i < len; ++i) {
    lookup[i] = code[i]
    revLookup[code.charCodeAt(i)] = i
  }

  revLookup['-'.charCodeAt(0)] = 62
  revLookup['_'.charCodeAt(0)] = 63
}
// VVVVVVVV
init()
//^^^^^^
``` 
taking it out of the function and putting it at the top level doesn't fix it either only differing the init call

**Comment by Rich-Harris** [comment]
Tree-shaking that code away would involve some pretty sophisticated static analysis. My hope with #902 is that we'll be able to make some progress in that direction. It's too early to make promises, but it'd be a useful exercise to think through what kind of smarts Rollup would need to handle each of these cases. So:

```js
Object.defineProperty(Buffer, Symbol.species, {
  value: null,
  configurable: true
})
```

Rollup knows that, for example, `Object.keys(Buffer)` could be removed, because it knows that `Object.keys` [is a pure function](https://github.com/rollup/rollup/blob/master/src/utils/pureFunctions.js#L17). Because `Object.defineProperty` isn't in that list, it assumes the worst. If it understood that O.dP` *only* affects the first argument, then it could remove that statement because it knows that `Buffer` isn't used. That seems possible.

```js
Buffer.TYPED_ARRAY_SUPPORT = global.TYPED_ARRAY_SUPPORT !== undefined
  ? global.TYPED_ARRAY_SUPPORT
  : typedArraySupport()
function typedArraySupport () {
  try {
    var arr = new Uint8Array(1)
    arr.__proto__ = {__proto__: Uint8Array.prototype, foo: function () { return 42 }}
// VVVVVVVV
    return arr.foo() === 42 && // typed array instances can be augmented
        typeof arr.subarray === 'function' && // chrome 9-10 lack `subarray`
        arr.subarray(1, 1).byteLength === 0 // ie10 has broken `subarray`
// ^^^^^^^^^^^^^
  } catch (e) {
    return false
  }
}
```

As of #902, Rollup will know that `arr` could only have one value – `new Uint8Array(1)`. In order to know that `arr.subarray(...)` didn't have side-effects, it would need to understand prototype chains *and* know that `Uint8Array.prototype.subarray` could either be a side-effect-free method that returns another Uint8Array *or* `undefined`. In order to know that `arr.foo()` didn't have side-effects, it'd need to be smarter still – it would need to understand that you can override the prototype chain. That feels a little bit hairy (what if the `arr.__proto__` assignment happened inside a function? Could it still trust its knowledge of `Uint8Array.prototype`? Tricky.)

```js
function init () {
  var code = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/'
  for (var i = 0, len = code.length; i < len; ++i) {
    lookup[i] = code[i]
    revLookup[code.charCodeAt(i)] = i
  }

  revLookup['-'.charCodeAt(0)] = 62
  revLookup['_'.charCodeAt(0)] = 63
}
// VVVVVVVV
init()
//^^^^^^
```

We know that `code` is a string, so we can know that `code.charCodeAt(...)` is a side-effect-free function that returns a number. With that dealt with, Rollup should already recognise that it can disregard assignments to `lookup` and `revLookup` *unless* they're used elsewhere in the bundle.

So in summary, some of these things can be massaged away, some might be a little bit harder. I think smarter static analysis is the right direction to go in (rather than brute-forcing Rollup into ignoring certain modules, etc – we tried that once and it was a disaster) but we'll need to try and see how far we get.

**Comment by calvinmetcalf** [comment]
A shorter term solution might be the ability to annotate your code to tell
rollup it is side effect free similar to how Istanbul or eslint allow you

**Comment by GGAlanSmithee** [comment]
Another option would be to add a `pureFunctions` config, which would be a list of functions that you, as a user, consider to be pure. This could then be concatenated with the one in [pureFunctions.js](https://github.com/rollup/rollup/blob/master/src/utils/pureFunctions.js#L17) before parsing?

**Comment by calvinmetcalf** [comment]
@GGAlanSmithee that wouldn't necessarily be helpful as what we need is a way of saying that a function only modifies the certain arguments which would help with `Object.defineProperty` and other similar cases, but that would only fix the first example, the third one would be fixed by this pull as everything else was removed from the file, the second example is a bit more complex as it's, from the perspective of rollup, a user defined function which would be harder to name.

your solution would work for own code but it would also be tricky for dependencies in context this is coming up in my work trying to get the node standard library shims to work with rollup so it wouldn't be code directly used in your own app.

**Comment by GGAlanSmithee** [comment]
@calvinmetcalf Ah, I see. Mabye that could help with `Object.assign` too then, which is also considered to be impure. (because it invokes getters/setters?)

**Comment by calvinmetcalf** [comment]
@Rich-Harris your idea for handling functions that only modify the first thing could work with 

'Object.defineProperty',
	'Object.defineProperties',
	'Object.assign',
	'Object.freeze',
	'Object.preventExtensions',
	'Object.seal',
	'Object.setPrototypeOf'

and fix a whole lot of subtle side effect bugs

**Comment by bfred-it** [comment]
Can this be rescued or should it be closed?

**Comment by calvinmetcalf** [comment]
this has been bit rotted something fierce so I'll close it
